packages/rs-drive-abci/src/abci/handler/mod.rs

```diff
@@ -208,33 +218,41 @@ where
 
         let block_execution_outcome::v0::BlockExecutionOutcome {
             app_hash,
-            tx_results,
+            state_transition_results,
             validator_set_update,
         } = run_result.into_data().map_err(Error::Protocol)?;
 
         // We need to let Tenderdash know about the transactions we should remove from execution
-        let (tx_results, tx_records): (Vec<ExecTxResult>, Vec<TxRecord>) = tx_results
+        let tx_results_and_records = state_transition_results
             .into_iter()
             .map(|(tx, result)| {
-                if result.code > 0 {
+                let tx_result = result.try_into_tx_result(platform_version)?;
+
+                let tx_record = if tx_result.code > 0 {
                     (
-                        result,
+                        tx_result,
                         TxRecord {
                             action: TxAction::Removed as i32,
                             tx,
                         },
                     )
                 } else {
                     (
-                        result,
+                        tx_result,
                         TxRecord {
                             action: TxAction::Unmodified as i32,
                             tx,
                         },
                     )
-                }
+                };
+
+                Ok(tx_record)
             })
-            .unzip();
+            .collect::<Result<Vec<(ExecTxResult, TxRecord)>, Error>>()
+            .map_err(proto::ResponseException::from)?;
+
+        let (tx_results, tx_records): (Vec<ExecTxResult>, Vec<TxRecord>) =
+            tx_results_and_records.into_iter().unzip();
```

Comment:
why remove unzip? This is less efficient.

Reply:
We need to collect result first and then we can do unzip.

Reply:
why?

Reply:
We do `map` first because we need to create `ExecTxResult` and `TxRecord` from tx and state transition result (we need both for `TxRecord`). Then we use `try_into_platform_versioned` that returns `Result`. So map should return result. We can't unzip result so we need to handle result. I can change it to the `for` loop.